improver/utilities/field_texture.py

```diff
@@ -0,0 +1,208 @@
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Module containing plugin to calculate the texture in a field considering edge transitions."""
+
+import iris
+import numpy as np
+from iris.exceptions import CoordinateNotFoundError
+
+from improver import BasePlugin
+from improver.metadata.constants.attributes import MANDATORY_ATTRIBUTE_DEFAULTS
+from improver.metadata.utilities import create_new_diagnostic_cube
+from improver.nbhood.square_kernel import SquareNeighbourhood
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_manipulation import collapsed
+
+
+class FieldTexture(BasePlugin):
+    """Plugin to calculate the texture in a binary field considering edge transitions
+
+    1) Takes a binary field that has been thresholded and looks for the transitions/edges
+       in the field that mark out a transition.
+    2) The transition calculation is then fed into the neighbourhooding code
+       (_calculate_ratio) to calculate a ratio for each cell.
+    3) The new cube of ratios is then thresholded and the realization coordinates
+       are collapsed to generate a mean of the thresholded ratios.
+    """
+
+    def __init__(self, nbhood_radius, ratio_threshold):
+        """
+
+        Args:
+            nbhood_radius (float):
+                The neighbourhood radius in metres within which the number of potential
+                transitions should be calculated. This forms the denominator in the
+                calculation of the ratio of actual to potential transitions that indicates a
+                field's texture. A larger radius should be used for diagnosing larger-scale
+                textural features.
+
+            ratio_threshold (float):
+                A unit-less threshold value that defines the ratio value above which
+                the field is considered clumpy and below which the field is considered
+                more contiguous.
+
+        """
+        self.nbhood_radius = nbhood_radius
+        self.ratio_threshold = ratio_threshold
+
+    def _calculate_ratio(self, cube, radius):
+        """
+        Calculates the ratio of actual to potential value transitions in a
+        neighbourhood about each cell.
+
+        The process is as follows:
+
+            1. For each grid cell find the number of cells of value 1 in a surrounding
+               neighbourhood of a size defined by the arg radius. The potential
+               transitions within that neighbourhood are defined as the number of
+               orthogonal neighbours (up, down, left, right) about cells of value 1.
+               This is 4 times the number of cells of value 1.
```

Comment:
This isn't actually what this function is doing.  The square neighbourhood, I believe, also looks at the four corners - meaning you are looking at all 8 surrounding grid cells in this code.  @bayliffe is this the desired behaviour?

Reply:
It does exactly what it says Caroline. It looks for all the values in a neighbourhood that have a value of 1, that is all the neighbourhooding code does. You're right that this is any point in the neighbourhood regardless of its position relative to the central grid cell, but that is what we want. Each of the cells of value 1 could have "downhill" transitions to neighbours of value 0. We multiply the number of cells of value 1 in the neighbourhood by 4 (rather than 8) to total up the potential number of orthogonal neighbour transitions that might exist in the whole neighbourhood.

Reply:
Yep OK.  Got it now.  Thanks.